activedirectory/azure-ad-integration-spring-boot-autoconfigure/src/main/java/com/microsoft/azure/autoconfigure/adintegration/AzureADJwtTokenFilter.java

```diff
@@ -0,0 +1,118 @@
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class AzureADJwtTokenFilter extends OncePerRequestFilter {
+    private static final Logger log = LoggerFactory.getLogger(AzureADJwtTokenFilter.class);
+
+    private static final String TOKEN_HEADER = "Authorization";
+    private static final String TOKEN_TYPE = "Bearer ";
+
+    private AzureADJwtFilterProperties aadJwtFilterProp;
+
+    public AzureADJwtTokenFilter(AzureADJwtFilterProperties aadJwtFilterProp) {
+        this.aadJwtFilterProp = aadJwtFilterProp;
+    }
+
+    private AuthenticationResult acquireTokenForGraphApi(
+            String tokenEncoded,
+            String tenantId) throws Throwable {
+        ClientCredential credential = new ClientCredential(aadJwtFilterProp.getClientId(), aadJwtFilterProp.getClientSecret());
+        ClientAssertion assertion = new ClientAssertion(tokenEncoded);
+
+        AuthenticationResult result = null;
+        ExecutorService service = null;
+        try {
+            service = Executors.newFixedThreadPool(1);
+            AuthenticationContext context = new AuthenticationContext(
+                    aadJwtFilterProp.getAadSignInUri()+ tenantId + "/",
+                    true,
+                    service);
+            Future<AuthenticationResult> future = context
+                    .acquireToken(aadJwtFilterProp.getAadGraphAPIUri(), assertion, credential, null);
+            result = future.get();
+        } catch (ExecutionException e) {
+            throw e.getCause();
+        } finally {
+            service.shutdown();
+        }
+
+        if (result == null) {
+            throw new ServiceUnavailableException("authentication result was null");
+        }
+        return result;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+
+        String authHeader = request.getHeader(TOKEN_HEADER);
+
+        if (authHeader != null && authHeader.startsWith(TOKEN_TYPE)) {
+            try {
+                String tokenEncoded = authHeader.replace(TOKEN_TYPE, "");
+                AzureADJwtToken jwtToken = new AzureADJwtToken(tokenEncoded);
+
+                AzureADUserProfile userProfile;
+                List<GrantedAuthority> authorities = new ArrayList<GrantedAuthority>();
+                try {
+                    String tid = jwtToken.getClaim("tid").toString();
+
+                    AuthenticationResult result = acquireTokenForGraphApi(
+                            tokenEncoded,
+                            tid);
+                    userProfile = new AzureADUserProfile(result.getAccessToken());
```

Comment:
looks like authz code will always run (which will retrieve user's group memberships). But this is not really necessary unless authZ requirements are explicitly configred on the resources that this filter is protecting